controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go

```diff
@@ -259,6 +262,55 @@ func (in *KubeadmControlPlane) validateCoreDNS() (allErrs field.ErrorList) {
 	return allErrs
 }
 
+func (in *KubeadmControlPlane) validateCoreDNSVersion(prev *KubeadmControlPlane) (allErrs field.ErrorList) {
+	if in.Spec.KubeadmConfigSpec.ClusterConfiguration == nil || prev.Spec.KubeadmConfigSpec.ClusterConfiguration == nil {
+		return allErrs
+	}
+	if prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag == "" {
+		return allErrs
+	}
+	dns := &in.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+	//return if the type is anything other than empty (default), or CoreDNS.
+	if dns.Type != "" && dns.Type != kubeadmv1.CoreDNS {
+		return allErrs
+	}
+
+	fromVersion, err := util.ParseMajorMinorPatch(prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.InternalError(
+				field.NewPath("spec", "kubeadmConfigSpec", "clusterConfiguration", "dns", "imageTag"),
+				fmt.Errorf("failed to parse CoreDNS current version: %v", prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag),
+			),
+		)
+		return allErrs
+	}
+
+	toVersion, err := util.ParseMajorMinorPatch(dns.ImageTag)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec", "kubeadmConfigSpec", "clusterConfiguration", "dns", "imageTag"),
+				dns.ImageTag,
+				fmt.Sprintf("failed to parse CoreDNS target version: %v", dns.ImageTag),
+			),
+		)
+		return allErrs
+	}
+
+	if err := migration.ValidUpMigration(fromVersion.String(), toVersion.String()); err != nil {
```

Comment:
Can we allow for the version major.minor.patch to be the same but the actual strings are different?

For example: `v0.1.6-foo.1` and `v0.1.6-foo.2` have the same version, but does the `ValidUpMigration` return an error in that case?

Also let's add a unit test that covers this case

Reply:
ValidUpMigration doesn't return an error in that case. https://github.com/kubernetes-sigs/cluster-api/pull/2707/files#diff-0d49cc99aec2899fcc03b149561ff8eeR305-R311 this test should cover that case.  It tries to update from `v1.6.6_foobar.1` to `v1.6.6_foobar.2`

Reply:
I see it's covered here https://github.com/coredns/corefile-migration/blob/master/migration/migrate.go#L424, great work :)

Reply:
For the case of the coredns migration, if the from to to version is the same, we just skip migrating the corefile:

https://github.com/coredns/corefile-migration/blob/master/migration/migrate.go#L129-L131